### PR TITLE
Hydra optuna sweeper custom search space 1909

### PR DESCRIPTION
**Pull request by MattiasDC**
## Motivation
This pull request addresses [issue 1906](https://github.com/facebookresearch/hydra/issues/1906). It introduces a new configuration option for the Hydra-Optuna-Sweeper plugin that allows extending Optuna trial objects via Pythonic search spaces.

For a project I would like to use Hydra and Optuna as a replacement of current customly written code. But in order to do that I need to have the possibility to do some complex trial configuration that is currently not supported by the Optuna plugin. With this extension we can easily combine the strength of Hydra with Optuna.

### Have you read the [Contributing Guidelines on pull requests]
Yes

## Test Plan
I've added unit tests and added an extra example that shows how to use this added functionality.

## Related Issues and PRs
 [issue 1906](https://github.com/facebookresearch/hydra/issues/1906)

### CODE REVIEW

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:228]: Setting fixed parameters as user_attrs in Optuna was previously not done. I think this however makes it more clear for Optuna that this parameter exists and future runs with it not being fixed, might use reuse the information of runs.

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:124]: Not an expert on imports, but I hope this doesn't bring any security issues that we need to take into account (e.g. ../../../etc/passwd). Since we split on '.', I think this is safe?

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:123]: I had to use 'Any' in the end since 'Callable' could not be used due to linting errors of it not being fully specified. Writing Callable[[DictConfig, Trial], None] doesn't work as mypy cannot defer the type from the getattr call. Furthermore 'isinstance(function, Callable[[DictConfig, Trial], None])' is not yet supported in Python. Generics can't be used in 'isinstance'

**Comment by Jasha10** [comment]
Hi @MattiasDC, thank you very much for your contribution! We will take a look at it shortly :)

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:123]: > 'isinstance(function, Callable[[DictConfig, Trial], None])' is not yet supported...

FYI there is a builtin [`callable`](https://docs.python.org/3/library/functions.html#callable) function that can do part of the job :)

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:210]: I think this line can be deleted.
```suggestion
```

**Review by Jasha10 — COMMENTED** [review]
A few initial comments:

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:123]: > 'isinstance(function, Callable[[DictConfig, Trial], None])' is not yet supported...

FYI there is a builtin callable function that can do part of the job :)

Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py]: Thanks for the info! Indeed, I could reused the get_method :)

**Comment by Jasha10** [comment]
Thanks for updating the PR, @MattiasDC.
@jieru-hu and I will need a few days to review this.

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:303]:
```suggestion
## EXPERIMENTAL:  Custom-Search-Space Optimization
```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:341]:
```suggestion
The `custom_search_space` configuration option refers to the `configure` function in `example/custom-search-space-objective.py`. This method accepts both a DictConfig with already set options and the trial object which needs further configuration. In this example we limit `z` the difference between `x` and `z` to be no more than 0.5.
Note that this `custom_search_space` API should be considered experimental and is subject to change.
```

**Review by Jasha10 — CHANGES_REQUESTED** [review]
We might want to change this API in the future; let's call it "experimental" for now.

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:212]: This change has caused CI to fail. I can reproduce the issue locally.
The problem is that we might have e.g. `name == "~z", as was the case with the failing test. In that case the f-string `f"++{name}"` would be equal to `"++~z"`, which is no bueno.

I am not convinced that adding `++` is the right move here.

**Review by Jasha10 — CHANGES_REQUESTED** [review]

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py]: I changed it back to remove '++', however I also extended the test to cover a case that I feel that must be supported (you can't define all parameters in advance in the DictConfig). Note that in the committed change 'w' cannot be overridden, because it's not known.

Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py]: Any ideas on how to resolve the currently failing test?

File: plugins/hydra_optuna_sweeper/news/1906.feature
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/news/1906.feature:1]:
```suggestion
Add experimental 'custom_search_space' configuration node to allow extending trial objects programmatically.
```

I think the word "option" could be confusing because it has a different meaning in the context of Hydra's defaults list (see e.g. ["Terminology - Config Group Option"](https://hydra.cc/docs/advanced/terminology/#config-group-option)).

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py]: I will take a look...

Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py]: @Jasha10 I've adapted the code to only add a '+' if the custom search space introduced a new parameter. Instead of by default add '++' to all parameters

**Comment by MattiasDC** [comment]
Hi @Jasha10,
Not trying to push you here, but I'm wondering whether you can give me an ETA? :)

**Comment by Jasha10** [comment]
Hi @MattiasDC, thanks for the nudge :)

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:224]: I am still not comfortable with this; it breaks consistency with Hydra's command-line API as well as with the precedent from PR #1886.
Hydra should fail fast if someone tries to set a parameter that does not exist in the config ... *unless* if the user explicitly adds a plus `+` to signal that adding the new parameter is intended.

This being said, I agree with you that it should be possible to use `custom_search_space` to add a new parameter to the config. Here is my proposal:
The user's custom search space function should call `trial.suggest_float` using a plus symbol for the name of the suggested parameter. For example:
```python
# custom-search-space-objective.py

...

def configure(cfg: DictConfig, trial: Trial) -> None:
    x_value = trial.params["x"]
    trial.suggest_float(
        "z",
        x_value - cfg.max_z_difference_from_x,
        x_value + cfg.max_z_difference_from_x,
    )
    trial.suggest_float("+w", 0.0, 1.0)  # note +w here, not w

...
```

**Review by Jasha10 — CHANGES_REQUESTED** [review]

**Comment by Jasha10** [comment]
Overall this PR looks good. Could you please rebase against Hydra's `main` branch to resolve the merge conflict?

**Comment by MattiasDC** [comment]
Hi @Jasha10 ,
I've addressed your remark and I've merged from main and resolved the conflicts with the new style params.
I also simplified it a bit further.
While doing the rebase I've messed up, so the commits might seem 'new' again as I had to do a force push.
My apologies for this.

I have 4 'failing' tests, but these seem to be flaky, is this a known problem?
They are related to jupyter_notebook and brew environment

**Comment by Jasha10** [comment]
Thanks Mattias!

> I have 6 'failing' tests, but these seem to be flaky, is this a known problem? They are related to jupyter_notebook and brew environment

Yes, the jupyter notebook failures are a known issue that motivated PR #2123. Incorporating that diff into this branch will (hopefully) solve the issue.

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:134]: In `OptunaSweeperImpl.__init__`, the `search_space` variable has type `Optional[DictConfig]`, but in this function `search_space` has type `Dict[str, BaseDistribution]`. This type mismatch is confusing... Could you please rename `search_space` here to a different name?

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:230]: Same remark as before about renaming `search_space`.

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:248]: Can this error be triggered by the user making a mistake with their config? If so, this should be a `ValueError` instead of an `AssertionError`.

File: plugins/hydra_optuna_sweeper/example/custom-search-space/config.yaml
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/example/custom-search-space/config.yaml:19]:
```suggestion
    params:
      x: range(-5.5, 5.5, 0.5)
      y: choice(-5, 0, 5)
    # `custom_search_space` should be a dotpath pointing to a
    # callable that provides search-space configuration logic:
    custom_search_space: .custom-search-space-objective.configure

```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:265]:
```suggestion
```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:286]:
```suggestion
    params:
      x: range(-5.5, 5.5, 0.5)
      y: choice(-5, 0, 5)
    # `custom_search_space` should be a dotpath pointing to a
    # callable that provides search-space configuration logic:
    custom_search_space: .custom-search-space-objective.configure
```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:293]:
````suggestion
```python
# example/custom-search-space-objective.py

...

def configure(cfg: DictConfig, trial: Trial) -> None:
    x_value = trial.params["x"]
    trial.suggest_float(
        "z",
        x_value - cfg.max_z_difference_from_x,
        x_value + cfg.max_z_difference_from_x,
    )
    trial.suggest_float("+w", 0.0, 1.0)  # note +w here, not w as w is a new parameter

...
```

````

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:258]:
```suggestion
You can find a full example in the same directory as before, where `example/custom-search-space-objective.py` implements a benchmark function to be minimized.
The example shows the use of Optuna's [pythonic search spaces](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html) in combination with Hydra.
Part of the search space configuration is defined in config files, and part of it is written in Python.
```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:260]:
```suggestion
<details><summary>Example: Custom search space configuration</summary>
```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:257]:
```suggestion
## EXPERIMENTAL:  Custom-Search-Space Optimization

Hydra's Optuna Sweeper allows users to provide a hook for custom search space configuration.
This means you can work directly with the `optuna.trial.Trial` object to suggest parameters.
To use this feature, define a python function with signature `Callable[[DictConfig, optuna.trial.Trial], None]`
and set the `hydra.sweeper.custom_search_space` key in your config to target that function.

```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:296]:
```suggestion
</details>

The method that `custom_search_space` points to must accepts both a DictConfig with already set options and a trial object which needs further configuration. In this example we limit `z` the difference between `x` and `z` to be no more than 0.5.
Note that this `custom_search_space` API should be considered experimental and is subject to change.
```

File: website/docs/plugins/optuna_sweeper.md
Comment (Jasha10) [inline website/docs/plugins/optuna_sweeper.md:300]:
```suggestion
  - search space parameters are set from the `hydra.sweeper.params` key in the config
```

**Review by Jasha10 — CHANGES_REQUESTED** [review]
The implementation looks good.
See my comments below:

File: plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:256]:
```suggestion
        abs(returns["best_params"]["x"] - returns["best_params"]["z"])
```

File: plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py
Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py]: Thanks, learned something new today! :)

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:249]: Yes it can, I will change this

**Comment by MattiasDC** [comment]
> Thanks Mattias!
> 
> > I have 6 'failing' tests, but these seem to be flaky, is this a known problem? They are related to jupyter_notebook and brew environment
> 
> Yes, the jupyter notebook failures are a known issue that motivated the recent PR #2123. Incorporating that diff into this branch will (hopefully) solve the issue.

Jupyter issues seem to be fixed. Brew issues are not. Is there any way to rerun the checks?

**Comment by Jasha10** [comment]
I've just triggered a rerun.
> Is there any way to rerun the checks?

You can always push an empty commit to the PR branch :P

**Comment by MattiasDC** [comment]
> I've just triggered a rerun.
> 
> > Is there any way to rerun the checks?
> 
> You can always push an empty commit to the PR branch :P

Seemed to do the trick!

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:135]: Changed it to search_space_distributions

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py]: Changed it to search_space_distributions

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (MattiasDC) [reply plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py]: Changed it accordingly

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:135]:
```suggestion
    search_space_distributions = dict()
```

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:141]:
```suggestion
            search_space_distributions[param_name] = value
```

File: plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py
Comment (Jasha10) [inline plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:144]:
```suggestion
    return search_space_distributions, fixed_params
```

**Review by Jasha10 — CHANGES_REQUESTED** [review]

**Review by Jasha10 — APPROVED** [review]
Looks good to me!

**Comment by MattiasDC** [comment]
Hi @Jasha10,

Thanks for your time and patience.
This was my first open source pull request, so if you have any feedback, feel free to leave some! :)

**Comment by MattiasDC** [comment]
Hydra Optuna sweeper plugin now allows to extend the trial object programmatically using a custom search space configuration option.

**Comment by MattiasDC** [comment]
Also, it seems I don't have write access to merge this PR, so I leave that up to you @Jasha10

**Comment by Jasha10** [comment]
> Hi @Jasha10,
> 
> Thanks for your time and patience. This was my first open source pull request, so if you have any feedback, feel free to leave some! :)

Wow, well you did a nice job with it! It was good collaborating with you 😄

@jieru-hu, could you please give this a final review before we merge?

**Comment by jieru-hu** [comment]
looks really good! thanks @MattiasDC !

**Review by jieru-hu — APPROVED** [review]

**Pull request merged**
